calcom/cal.com
# respect local set timezone and update url on mismatch

## What does this PR do?

### 1. will keep `timeZone()` and the offset from url in sync. 

If URL offset does not match local set `timeZone()`, the URL will be fixed. However if we want, we can  
- either maintain a list of offsets to timezones and set the `timeZone()`
- add another query param representing the timezone (this would be a breaking change tho)

### 2. Ensure the local selected timezone will make its way to the URL
If no date is selected, the URL contains no offset. By selecting the first date, the local set timezone (`timeZone()`) will be used.

Fixes [#2482](https://github.com/calcom/cal.com/issues/2482)

## Type of change




## How should this be tested?




1. Go to booking page
2. Select a time zone different from the default before selecting a date.
3. Select a date.

=> now the offset from the picker and the offset in the URL should match. Also the timeslots should be in the right timezone.


1. Go to booking page
2. Select a time zone
3. Select a date
4. Modify the URL (eg: `http://localhost:3000/free/30min?date=2022-04-18%2B0300` -> `http://localhost:3000/free/30min?date=2022-04-18%2B0100`)

=> the effect should trigger and the URL will be changed back (`http://localhost:3000/free/30min?date=2022-04-18%2B0300`)

=> now the offset from the picker and the offset in the URL should match. Also the timeslots should be in the right timezone.
